packages/@aws-cdk/aws-lambda/lib/function-hash.ts

```diff
@@ -1,21 +1,39 @@
 import * as crypto from 'crypto';
 import { CfnResource, Stack } from '@aws-cdk/core';
 import { Function as LambdaFunction } from './function';
+import { sortedStringify } from './private/sort';
 
-export function calculateFunctionHash(fn: LambdaFunction) {
+export const ENV_SALT = '4b2bff42-e517-41d1-9b5f-92aacf41842b';
+
+export function calculateFunctionHash(hashAlgorithm: 'v1' | 'v2', fn: LambdaFunction) {
   const stack = Stack.of(fn);
 
   const functionResource = fn.node.defaultChild as CfnResource;
 
   // render the cloudformation resource from this function
   const config = stack.resolve((functionResource as any)._toCloudFormation());
-
   const hash = crypto.createHash('md5');
-  hash.update(JSON.stringify(config));
+
+  if (hashAlgorithm === 'v2') {
+    hash.update(sortedStringify(usefulKeys(config)));
+  } else {
+    hash.update(JSON.stringify(config));
+  }
 
   return hash.digest('hex');
 }
 
+function usefulKeys(cfnObject: any): any {
```

Comment:
Feels like this function deserves some docs to point to source of what should be included/removed. Is https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionConfiguration.html the best page?

Also, anything you can think of to add some form of validation when a new property is added to Lambda to check if it should be included or not? We could have an explicit allow- and deny-list for each property that must be updated whenever the cfnspec is changed, but that does feel a bit like overkill. I'm worried though about a new property added that shouldn't be included in the hash (but is), and then we need to create a 'v3', a new `ENV_SALT` and have another one-time logical ID update to fix it.

Reply:
Slight correction: we won't need a 'v3'. 

We'll just change the salt and invalidate all customer function versions. I suppose that's just as bad.

Reply:
If an authoritative source for this list exists and the expectation is to just look up and keep in sync, this makes total sense.

I'm wondering how much we'll get it right even with an allow/deny list if a source of truth doesn't exist and new properties are usually added by the community.

Reply:
Feels like the mode where we SHOULD have created a new version, but didn't, is better to have as the fail case than the reverse.

It's possible to work around the former (by changing the ID yourself), whereas it's hard to work around the latter.

Based on that, isn't an allowlist the safer option?

Reply:
Added an allowlist.